*Filière :* ING3 - Tronc commun

*Auteurs :*
**Michael KISUKA** et
**Nivethan SIVANESAN**

*Professeur référent :* 
**TAJINI Badr**

*Année :* 2024-2025

**ESIEE-IT**


# Phase 3 : Construction et Entraînement d'un Modèle MLP

Dans cette phase, nous allons construire un modèle de réseau de neurones multicouches (MLP) en utilisant **scikit-learn**.  
Nous allons ensuite l'entraîner sur le dataset **MNIST** et évaluer ses performances.

### Objectifs :
1. Charger et prétraiter les données MNIST.
2. Construire un modèle MLP simple avec scikit-learn.
3. Entraîner le modèle et mesurer sa précision.
4. Observer quelques prédictions effectuées par le modèle.


In [1]:
from sklearn.datasets import fetch_openml
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import numpy as np

# 1. Charger le dataset MNIST
mnist = fetch_openml('mnist_784', version=1)
X = mnist.data.astype(np.float32) / 255.0  # Normaliser les données
y = mnist.target.astype(int)

# 2. Diviser les données en ensembles d'entraînement et de test
#    C'est crucial pour évaluer les performances du modèle sur des données qu'il n'a jamais vues.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 3. Construire un modèle MLP (vous pouvez utiliser un des modèles de l'étape 2 ou en créer un nouveau)
#    Commençons par un modèle simple pour l'entraînement initial.
mlp = MLPClassifier(hidden_layer_sizes=(50,), max_iter=10, random_state=42)
print("Modèle MLP créé :", mlp)

# 4. Entraîner le modèle sur les données d'entraînement
#    C'est là que le modèle apprend à reconnaître les chiffres.
print("\nDébut de l'entraînement du modèle...")
mlp.fit(X_train, y_train)
print("Entraînement terminé.")

# 5. Faire des prédictions sur l'ensemble de test
#    Utiliser le modèle entraîné pour prédire les chiffres sur l'ensemble de test.
y_pred = mlp.predict(X_test)

# 6. Évaluer les performances du modèle en calculant la précision
#    La précision mesure le pourcentage de chiffres correctement classifiés.
accuracy = accuracy_score(y_test, y_pred)
print(f"\nPrécision du modèle sur l'ensemble de test : {accuracy * 100:.2f}%")

# 7. (Facultatif) Afficher quelques prédictions et les étiquettes réelles pour comparaison
print("\nQuelques prédictions et étiquettes réelles :")
for i in range(10):
    print(f"Image {i+1}: Prédiction = {y_pred[i]}, Réel = {y_test.iloc[i]}")

Modèle MLP créé : MLPClassifier(hidden_layer_sizes=(50,), max_iter=10, random_state=42)

Début de l'entraînement du modèle...
Entraînement terminé.

Précision du modèle sur l'ensemble de test : 96.46%

Quelques prédictions et étiquettes réelles :
Image 1: Prédiction = 8, Réel = 8
Image 2: Prédiction = 4, Réel = 4
Image 3: Prédiction = 8, Réel = 8
Image 4: Prédiction = 7, Réel = 7
Image 5: Prédiction = 7, Réel = 7
Image 6: Prédiction = 0, Réel = 0
Image 7: Prédiction = 6, Réel = 6
Image 8: Prédiction = 2, Réel = 2
Image 9: Prédiction = 7, Réel = 7
Image 10: Prédiction = 4, Réel = 4


c:\Users\micha\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (10) reached and the optimization hasn't converged yet.
  warnings.warn(


On peut voir que le processus prend un certain temps à se mettre en marche car il prend le temps de traiter toutes les images. On voit que sur l'ensemble du test, la précision du modèle est à peu près à 96.46% ce qui est très proche de 100%.

Si on s'attarde sur les résultats, on se rend compte que pour chaque image, notre modèle propose une prédication sous forme de chiffre et qu'au final lors du resultat final, on voit que la prédication était vrai et que la logique de notre modèle à été confirmé.

In [1]:
from sklearn.datasets import fetch_openml
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import numpy as np

# 1. Charger le dataset MNIST
mnist = fetch_openml('mnist_784', version=1)
X = mnist.data.astype(np.float32) / 255.0  # Normaliser les données
y = mnist.target.astype(int)

# 2. Diviser les données en ensembles d'entraînement et de test
#    C'est crucial pour évaluer les performances du modèle sur des données qu'il n'a jamais vues.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 3. Construire un modèle MLP (vous pouvez utiliser un des modèles de l'étape 2 ou en créer un nouveau)
#    Commençons par un modèle simple pour l'entraînement initial.
mlp = MLPClassifier(hidden_layer_sizes=(100, 50), max_iter=50, random_state=42)
print("Modèle MLP créé :", mlp)

# 4. Entraîner le modèle sur les données d'entraînement
#    C'est là que le modèle apprend à reconnaître les chiffres.
print("\nDébut de l'entraînement du modèle...")
mlp.fit(X_train, y_train)
print("Entraînement terminé.")

# 5. Faire des prédictions sur l'ensemble de test
#    Utiliser le modèle entraîné pour prédire les chiffres sur l'ensemble de test.
y_pred = mlp.predict(X_test)

# 6. Évaluer les performances du modèle en calculant la précision
#    La précision mesure le pourcentage de chiffres correctement classifiés.
accuracy = accuracy_score(y_test, y_pred)
print(f"\nPrécision du modèle sur l'ensemble de test : {accuracy * 100:.2f}%")

# 7. (Facultatif) Afficher quelques prédictions et les étiquettes réelles pour comparaison
print("\nQuelques prédictions et étiquettes réelles :")
for i in range(10):
    print(f"Image {i+1}: Prédiction = {y_pred[i]}, Réel = {y_test.iloc[i]}")

Modèle MLP créé : MLPClassifier(hidden_layer_sizes=(100, 50), max_iter=50, random_state=42)

Début de l'entraînement du modèle...
Entraînement terminé.

Précision du modèle sur l'ensemble de test : 97.55%

Quelques prédictions et étiquettes réelles :
Image 1: Prédiction = 8, Réel = 8
Image 2: Prédiction = 4, Réel = 4
Image 3: Prédiction = 8, Réel = 8
Image 4: Prédiction = 7, Réel = 7
Image 5: Prédiction = 7, Réel = 7
Image 6: Prédiction = 0, Réel = 0
Image 7: Prédiction = 6, Réel = 6
Image 8: Prédiction = 2, Réel = 2
Image 9: Prédiction = 7, Réel = 7
Image 10: Prédiction = 4, Réel = 4


Pour améliorer la précision du modèle, deux modifications principales peuvent être apportées. La première consiste à augmenter le nombre d'itérations d'entraînement en ajustant le paramètre max_iter. En augmentant cette valeur, le modèle a plus de temps pour apprendre les motifs dans les données, ce qui peut améliorer la précision. Par exemple, en passant de max_iter=10 à max_iter=50 ou max_iter=100, on peut observer une meilleure performance. Cependant, il faut faire attention au surapprentissage, où le modèle mémorise trop les données d'entraînement et devient moins efficace sur des données nouvelles.

Modifier l'architecture du réseau en augmentant le nombre de neurones ou en ajoutant des couches cachées (hidden_layer_sizes) permet d'affiner la capacité du modèle à reconnaître des motifs complexes. Une architecture plus profonde améliore les performances, mais une complexité excessive peut nuire à la généralisation. L'objectif est de trouver un équilibre entre précision et efficacité.


### Conclusion

Les résultats avant et après la modification du code montrent une amélioration notable dans la performance du modèle de classification MLP appliqué aux données MNIST. Dans la version initiale, le modèle utilise un réseau de neurones avec une seule couche cachée de 50 neurones et un nombre d'itérations limité à 10, ce qui limite la capacité d'apprentissage et, par conséquent, la précision du modèle. Après modification, l'architecture du modèle a été ajustée pour inclure une couche cachée de 100 neurones, suivie d'une autre de 50, et le nombre d'itérations a été augmenté à 50. Ces ajustements permettent au modèle d'apprendre de manière plus approfondie, ce qui se reflète dans une meilleure précision sur l'ensemble de test. En somme, la modification du code permet au modèle d'atteindre des performances accrues en offrant une plus grande capacité d'apprentissage et un entraînement plus long.